src/common/framework/preprocessor.ts

```diff
@@ -0,0 +1,101 @@
+import { assert } from './util/util.js';
+
+// The state of the preprocessor is a stack of States.
+// The transitions in the state space are:
+// - Sibling elif
+// - Sibling else
+// - Sibling endif
+// - Child if
+const enum State {
+  Seeking, // Still looking for a passing condition
+  Passing, // Currently inside a passing condition
+  Skipping, // Have already seen a passing condition; now skipping the rest
+}
+
+export function pp(parts: TemplateStringsArray, ...predicates: readonly boolean[]): string {
+  let result = '';
+
+  const stateStack: State[] = [State.Passing];
+
+  for (let partIndex = 0; partIndex < parts.length; ++partIndex) {
+    const lines = parts[partIndex].split('\n');
+
+    checkDirectiveComment(partIndex, lines[0]);
```

Comment:
I don't understand the motivation having this requirement:

The first line has to be empty, or only spaces, or only spaces followed by a #-initiated comment.

Maybe I'm not reading that correctly?

Reply:
That's correct, and I admit it's kind of silly. I initially did it mainly because it simplified the implementation: the first line of the first `part`*s isn't treated differently from the first line of other `part`s.

\* where `part` is e.g. `"# c1\nA\n#if "` or `" # c2\nB\n#endif"` as parsed by JS in:
```js
pp` # c1
A
#if ${true} # c2
B
#endif`
```

The line with `# c1` can be treated like it's a directive-tail comment just like `# c2`.

Also because it forces a style where there's no code on the first line of an expression that looks like:

```js
    const x = pp`line 1 <- code here wouldn't line up with the rest of the code
line 2 <- lines up
line 3
`;
```

Though maybe it would be nice to allow a style like this. I could easily use the number of spaces at the beginning of the second line to decide the indentation of the directives
```
    const y = pp`
      line 2
      line 3
      %if <- this has spaces before it, so currently treated like plain text, not as a directive
    `;
```